x/oracle/keeper/abci.go

```diff
@@ -6,164 +6,197 @@ import (
+		validatorVotesMap[validatorAddr.String()] = true
 
-			// find total voting votedPower
-			votedPower += k.stakingKeeper.Validator(ctx, sdk.ValAddress(val)).GetConsensusPower()
+		// remove the miss counter in the current voting window for validators who have already voted
+		k.DeleteMissCounter(ctx, validatorAddr)
 
-			// save the oracle data for later processing
-			for _, oda := range msg.OracleData {
-				od, err := types.UnpackOracleData(oda)
-				if err != nil {
-					panic(err)
-				}
+		// find total voting votedPower
+		validator := k.stakingKeeper.Validator(ctx, sdk.ValAddress(validatorAddr))
+		if validator == nil {
+			// validator nor found, continue with next vote
+			return false
+		}
 
-				detailedMap[od.Type()][val.String()] = od
-				collectionMap[od.Type()] = append(collectionMap[od.Type()], od)
-			}
+		votedPower += validator.GetConsensusPower()
 
-			// delete the vote as we no longer require it
-			k.DeleteOracleDataVote(ctx, val)
+		// Safety check. Already validated on msg validation
+		if vote.Feed == nil || len(vote.Feed.OracleData) == 0 {
+			k.Logger(ctx).Debug("attempted to process empty oracle data from feed", "validator", validatorAddr.String())
 			return false
-		})
-
-		// iterate over the full list of validators to increment miss counters
-		totalPower := int64(0)
-		for _, val := range k.stakingKeeper.GetBondedValidatorsByPower(ctx) {
-			totalPower += val.GetConsensusPower()
-			valaddr := sdk.AccAddress(val.GetOperator())
-			if !contains(voted, valaddr.String()) {
-				k.IncrementMissCounter(ctx, valaddr)
+		}
+
+		// save the oracle data for later processing
+		for _, oracleDataAny := range vote.Feed.OracleData {
+			oracleData, err := types.UnpackOracleData(oracleDataAny)
+			if err != nil {
+				// NOTE: this should never panic as the oracle data had already been checked before
+				// setting it to store.
+				panic(err)
 			}
+
+			// add oracle data to maps
+			submittedFeedMap[oracleData.Type()][validatorAddr.String()] = oracleData
+
 		}
 
-		// if the voted_power/total_power > params.VoteThreshold then we store the averages in the store
-		storeAverages := sdk.NewDec(votedPower).Quo(sdk.NewDec(totalPower)).GT(params.VoteThreshold)
+		// delete the vote as it has already been processed
+		// TODO: consider keeping the votes for a few voting windows in order to
+		// be able to submit evidence of inaccurate data feed
+		k.DeleteOracleDataVote(ctx, validatorAddr)
+		return false
+	})
+
+	// iterate over the full list of validators to increment miss counters if they didn't vote
+	totalPower := int64(0)
+	k.stakingKeeper.IterateBondedValidatorsByPower(ctx, func(_ int64, validator stakingtypes.ValidatorI) bool {
+		totalPower += validator.GetConsensusPower()
+		validatorAddr := sdk.AccAddress(validator.GetOperator())
+
+		if !validatorVotesMap[validatorAddr.String()] {
+			// TODO: this is wrong because the feeder could submit a single oracle data uniswap pair instead of all the required
```

Comment:
The way I had this in the prior implemenation is that we required a # of different markets (i.e. 100) from each validator.